### ISSUE TITLE
build: allow filtering runtime environment

### ISSUE THREAD

**Issue by ysndr** [flox/flox#2065]
**Describe the feature:**

Add the control knobs in the manifest to control the runtime environment of produced packages.

```
build.<name> = {
  runtime-packages = Runtime
}

Runtime := null | [ STRING ]
```

By default `build.<package>.runtime-packages = null` the built package will only be allowed to reference packages in the `toplevel` group.
Alternatively a build can specify a list of install id **from the `toplevel` group** to reduce the closure even further.

**Describe a use case of this feature:**

Particularly for manifest builds, we would like to produce environments that only include a specific group, i.e. are guaranteed to depend on a single catalog page / revision of nixpkgs.

**Acceptance criteria:**
1. add `runtime-packages` attribute to `ManifestBuildDescriptor`
2. discuss validation of runtime packages; the runtime attribute can only reference packages from the `toplevel` group. At the moment it is unclear whether that is enforced during the build of a package, the build of the environment, or locking of the environment (or at another point altogether)

**Further work**
discuss changes to `flox-build.mk` calls & implement, possible to split out of

**Comment by limeytexan** [comment]
As discussed offline:

- need the ability to specify different runtime closures for different packages
- all of which should be comprised of packages from the toplevel package group only
- and there can be overlap between the packages included in the various runtime closures

So I'm guessing the runtime attribute needs to be an array of package "names" as found in the `[install]` block (i.e. the keys of that section).

**Comment by mkenigs** [comment]
I thought even if runtime was set we'd have to restrict to a single package group. Is that not the case?

**Comment by ysndr** [comment]
I guess that's a question of defaults and for simplicity it would be off by default under this planning but apart from it being the less frictious default I'm not having a strong opinion here.

**Comment by mkenigs** [comment]
I think I mistyped and meant unset 😬 I thought we have to restrict to a single package group no matter what, whether runtime is set or unset. In which case I would think runtime would be `null | [ STRING ]`

**Comment by mkenigs** [comment]
@ysndr can you just confirm exactly what we're settling on and update the description?

**Comment by ysndr** [comment]
My impression was that we _do_ want to support the current situation that is that the output can reference _any_ path from the environment (or the "toplevel" group), and encourage setting the runtime for smaller closures, compatibility, and apparently being able to rebuild with new nixpkgs (though i dont quite see why that's not possible just yet 🤔)

Other than that i think `runtime` or `runtime-packages` is a bit clearer than just packages since its more descriptive, but that's name bikeshedding in the end.

**Comment by limeytexan** [comment]
> My impression was that we do want to support the current situation that is that the output can reference any path from the environment (or the "toplevel" group)

Those are different things - the "toplevel" group yes, but anything in the environment outside of that group is a definite no. That's really the point - a package must be derived from and use packages from only a single nixpkgs revision.

> ... and encourage setting the runtime for smaller closures, compatibility, and apparently being able to rebuild with new nixpkgs (though i dont quite see why that's not possible just yet 🤔)

Yes, the `packages` attribute is all about being able to declare smaller runtime closures for built packages.

The requirement to restrict build dependencies to the "toplevel" group is what guarantees [ABI] compatibility down the line, minimizes closure size, and provides a strategy for rebuilding that package against successive nixpkgs revisions over time. The fact that builds can sometimes work using packages from a mix of nixpkgs revisions is not a reason to allow that, because when ABI incompatibilities do hit you in the face they are remarkably hard to detect/understand, and those are exactly the kinds of experiences that would lead even the most experienced developers to incorrectly conclude that flox was simply "buggy".

**Comment by limeytexan** [comment]
> Other than that i think runtime or runtime-packages is a bit clearer than just packages since its more descriptive, but that's name bikeshedding in the end.

Just noting FTR that Yannik and I discussed this offline and I'm happy with `runtime-packages` as implemented in #2257
